### ISSUE TITLE
Jetty server - writing to response before request is fully read gets blocked and times out.

### ISSUE THREAD

**Issue by flozano** [eclipse/jetty.project#11484]
**Jetty version(s)**
jetty 12.0.6

**Jetty Environment**
ee10

**Java version/vendor** `(use: java -version)`
```
openjdk 21.0.2 2024-01-16 LTS
OpenJDK Runtime Environment Zulu21.32+17-CA (build 21.0.2+13-LTS)
OpenJDK 64-Bit Server VM Zulu21.32+17-CA (build 21.0.2+13-LTS, mixed mode, sharing)
```
```
openjdk version "17.0.10" 2024-01-16
OpenJDK Runtime Environment Temurin-17.0.10+7 (build 17.0.10+7)
OpenJDK 64-Bit Server VM Temurin-17.0.10+7 (build 17.0.10+7, mixed mode)
```
and even
```
openjdk version "17.0.10" 2024-01-16
IBM Semeru Runtime Open Edition 17.0.10.0 (build 17.0.10+7)
Eclipse OpenJ9 VM 17.0.10.0 (build openj9-0.43.0, JRE 17 Mac OS X aarch64-64-Bit 20240116_514 (JIT enabled, AOT enabled)
OpenJ9   - 2c3d78b48
OMR      - ea8124dbc
JCL      - 2aad089841f based on jdk-17.0.10+7)
```

**OS type/version**
macOS sonoma 14.2, arm64
```
Darwin Franciscos-MacBook-Pro.local 23.2.0 Darwin Kernel Version 23.2.0: Wed Nov 15 21:53:18 PST 2023; root:xnu-10002.61.3~2/RELEASE_ARM64_T6000 arm64
```

**Description**

I have a servlet that looks like this:

```java
	protected void service(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
		LOGGER.info("Received request: {}", req.getRequestURI());
		resp.setContentType(req.getContentType());
		resp.setContentLength(req.getContentLength());
		resp.setCharacterEncoding(req.getCharacterEncoding());
		LOGGER.info("Copying input to output (bufferSize={})", bufferSize);
		resp.setStatus(200);
		try (var inputStream = req.getInputStream()) {
			try (var outputStream = resp.getOutputStream()) {
				if (bufferSize == 0) {
					var content = req.getInputStream().readAllBytes();
					resp.getOutputStream().write(content);
				} else {
					byte[] buffer = new byte[bufferSize];
					copyLarge(inputStream, outputStream, buffer);
				}
			}
		}
		LOGGER.info("Finished request: {}", req.getRequestURI());
	}
```
it's used to reproduce the issue, it serves no other purpose than to write the input as soon as it comes, with a configurable buffer size. 

The issue is that, depending of the copy buffer size + request size + jetty buffer response size combination, the request is "locked" and times-out.

I am not sure if this behavior is expected or even intended.

I guess if the condition is non-recoverable, jetty should write an immediate error response to the client if the request is not going to be fulfilled because of how buffers are, before the client fully writes the request (as it is allowed by RFC to write an error response before the entire request has been received), rather than letting the request timeout.

**How to reproduce?**

This is the reproducing project: https://github.com/flozano/jetty-issue/

**Comment by lachlan-roberts** [comment]
I tried your reproducer project and all the tests passed for me and I ran them multiple times.

Was this code working fine in an older version of Jetty and you experienced issues when you upgraded, if so which version of Jetty was it working with without failures?

Can you send the debug logs for one of your failing test cases?

**Comment by flozano** [comment]
Thanks for checking.

I am using this JVM on macOS (not sure if relevant):
```
❯ java --version
openjdk 21.0.2 2024-01-16 LTS
OpenJDK Runtime Environment Zulu21.32+17-CA (build 21.0.2+13-LTS)
OpenJDK 64-Bit Server VM Zulu21.32+17-CA (build 21.0.2+13-LTS, mixed mode, sharing)
```
<details><summary>and verified it fails for me for many combinations</summary>
<p>

```
❯ ./gradlew build

> Task :test

JettyIssue > test(int, int, int) > [11] 8192, 1024, 2000000 FAILED
    java.io.IOException at JettyIssue.java:45
        Caused by: java.io.IOException at Utils.java:391
            Caused by: java.io.IOException at SocketDispatcher.java:-2

JettyIssue > test(int, int, int) > [12] 8192, 1024, 4000000 FAILED
    java.io.IOException at JettyIssue.java:45
        Caused by: java.io.IOException at Utils.java:391
            Caused by: java.io.IOException at SocketDispatcher.java:-2

JettyIssue > test(int, int, int) > [15] 8192, 32768, 4000000 FAILED
    java.io.IOException at JettyIssue.java:45
        Caused by: java.io.IOException at Utils.java:391
            Caused by: java.io.IOException at SocketDispatcher.java:-2

JettyIssue > test(int, int, int) > [20] 1000000, 1024, 2000000 FAILED
    java.io.IOException at JettyIssue.java:45
        Caused by: java.io.IOException at Utils.java:391
            Caused by: java.io.IOException at SocketDispatcher.java:-2

JettyIssue > test(int, int, int) > [21] 1000000, 1024, 4000000 FAILED
    java.io.IOException at JettyIssue.java:45
        Caused by: java.io.IOException at Utils.java:391
            Caused by: java.io.IOException at SocketDispatcher.java:-2

JettyIssue > test(int, int, int) > [24] 1000000, 32768, 4000000 FAILED
    java.io.IOException at JettyIssue.java:45
        Caused by: java.io.IOException at Utils.java:391
            Caused by: java.io.IOException at SocketDispatcher.java:-2

JettyIssue > test(int, int, int) > [29] 4000000, 1024, 2000000 FAILED
    java.io.IOException at JettyIssue.java:45
        Caused by: java.io.IOException at Utils.java:391
            Caused by: java.io.IOException at SocketDispatcher.java:-2

JettyIssue > test(int, int, int) > [30] 4000000, 1024, 4000000 FAILED
    java.io.IOException at JettyIssue.java:45
        Caused by: java.io.IOException at Utils.java:391
            Caused by: java.io.IOException at SocketDispatcher.java:-2

JettyIssue > test(int, int, int) > [33] 4000000, 32768, 4000000 FAILED
    java.io.IOException at JettyIssue.java:45
        Caused by: java.io.IOException at Utils.java:391
            Caused by: java.io.IOException at SocketDispatcher.java:-2

36 tests completed, 9 failed

> Task :test FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///Users/flozano/Projects/java/demo/jetty-issue/build/reports/tests/test/index.html

* Try:
> Run with --scan to get full insights.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/8.6/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 1m 5s
4 actionable tasks: 2 executed, 2 up-to-date
```

</p>
</details> 

I will add debug logs for one of them in a moment.

**Comment by flozano** [comment]
Most failing tests fail in 5 seconds, like this one:
[failing-5-seconds.txt](https://github.com/jetty/jetty.project/files/14478105/failing-5-seconds.txt)


But this one fails in 10 seconds instead; not sure if it's a client-side artifact or if it could be relevant:
[failing-10-seconds.txt](https://github.com/jetty/jetty.project/files/14478109/failing-10-seconds.txt)

**Comment by flozano** [comment]
> Was this code working fine in an older version of Jetty and you experienced issues when you upgraded, if so which version of Jetty was it working with without failures?

No. This is new code. This dumb "mirror servlet" is something I am using to perform integration tests on a servlet filter.

**Comment by sbordet** [comment]
I also ran your reproducer on Linux and it always passes for me.

Looking at the logs, this is what happens:

```
08:43:17.766 [testing-jetty-21-316] DEBUG org.eclipse.jetty.io.SelectableChannelEndPoint -- Key interests updated 0 -> 4 on SocketChannelEndPoint@353fec5f[{l=/127.0.0.1:8089,r=/127.0.0.1:59833,OPEN,fill=-,flush=P,to=0/5000}{io=4/4,kio=4,kro=1}]->[HttpConnection@3fc7872f[p=HttpParser{s=CONTENT,556804 of 4000000},g=HttpGenerator@305248a0{s=COMMITTED}]=>HttpChannelState@4b754676{handling=Thread[#322,testing-jetty-21-322,5,main], handled=false, send=SENDING, completed=false, request=POST@0 http://127.0.0.1:8089/ HTTP/1.1}]
08:43:17.766 [testing-jetty-21-316] DEBUG org.eclipse.jetty.io.ManagedSelector -- updates 0
08:43:17.766 [testing-jetty-21-316] DEBUG org.eclipse.jetty.io.ManagedSelector -- Selector sun.nio.ch.KQueueSelectorImpl@1b3a9c81 waiting with 1 keys
08:43:22.723 [Scheduler-8197402-1] DEBUG org.eclipse.jetty.io.IdleTimeout -- SocketChannelEndPoint@353fec5f[{l=/127.0.0.1:8089,r=/127.0.0.1:59833,OPEN,fill=-,flush=P,to=4956/5000}{io=4/4,kio=4,kro=1}]->[HttpConnection@3fc7872f[p=HttpParser{s=CONTENT,556804 of 4000000},g=HttpGenerator@305248a0{s=COMMITTED}]=>HttpChannelState@4b754676{handling=Thread[#322,testing-jetty-21-322,5,main], handled=false, send=SENDING, completed=false, request=POST@0 http://127.0.0.1:8089/ HTTP/1.1}] idle timeout check, elapsed: 4956 ms, remaining: 44 ms
```

The server writes to the client, but the client does not read fast enough, and the TCP connection gets congested.
So the server stops writing and registers for `OP_WRITE` interest (`Key interests updated 0 -> 4`).

Then, nothing happens for almost 5 seconds, at which point there is a race between the scheduler thread to timeout the connection, and an NIO wakeup notifying that the TCP connection is writable.
In the case above, it just happens that the NIO wakeup wins (so you have an error after 10s), but in other cases the timeout wins (and you have an error after 5 seconds).

To sum up, the Jetty server is so fast writing the response, that the client cannot keep up.

Can you please try with Jetty's `HttpClient` rather than the JDK one, and see if you still get errors?
See https://eclipse.dev/jetty/documentation/jetty-12/programming-guide/index.html#pg-client-http-content-response.

**Comment by flozano** [comment]
Thanks again for checking!

I will modify the reproducer to use jetty client and see how it goes.

In the meantime, I had set the 5 second timeout to make the test run faster. This is what happens if I set to much higher (default if I recall correctly was 30, this is 45):
[failing-45-seconds.txt](https://github.com/jetty/jetty.project/files/14483218/failing-45-seconds.txt)


(BTW if this is not reproducing in Linux, maybe it's macOS specific?)

**Comment by sbordet** [comment]
Same problem with 45 seconds:

```
15:51:53.159 [testing-jetty-11-175] DEBUG org.eclipse.jetty.io.SelectableChannelEndPoint -- Key interests updated 0 -> 4 on SocketChannelEndPoint@69fd59ec[{l=/127.0.0.1:8089,r=/127.0.0.1:53002,OPEN,fill=-,flush=P,to=0/45000}{io=4/4,kio=4,kro=1}]->[HttpConnection@6305329c[p=HttpParser{s=CONTENT,556804 of 2000000},g=HttpGenerator@25b6bf8d{s=COMMITTED}]=>HttpChannelState@17eb662a{handling=Thread[#181,testing-jetty-11-181,5,main], handled=false, send=SENDING, completed=false, request=POST@0 http://127.0.0.1:8089/ HTTP/1.1}]
15:51:53.159 [testing-jetty-11-175] DEBUG org.eclipse.jetty.io.ManagedSelector -- updates 0
15:51:53.159 [testing-jetty-11-175] DEBUG org.eclipse.jetty.io.ManagedSelector -- Selector sun.nio.ch.KQueueSelectorImpl@5e0f5b9e waiting with 1 keys
15:52:38.099 [Scheduler-985487390-1] DEBUG org.eclipse.jetty.io.IdleTimeout -- SocketChannelEndPoint@69fd59ec[{l=/127.0.0.1:8089,r=/127.0.0.1:53002,OPEN,fill=-,flush=P,to=44939/45000}{io=4/4,kio=4,kro=1}]->[HttpConnection@6305329c[p=HttpParser{s=CONTENT,556804 of 2000000},g=HttpGenerator@25b6bf8d{s=COMMITTED}]=>HttpChannelState@17eb662a{handling=Thread[#181,testing-jetty-11-181,5,main], handled=false, send=SENDING, completed=false, request=POST@0 http://127.0.0.1:8089/ HTTP/1.1}] idle timeout check, elapsed: 44939 ms, remaining: 61 ms
```

After TCP congestion, 45 seconds elapsed and no events from NIO.

At this point really smells like a JVM bug to me, specific to MacOS.

Had you used the issue template, we would have had an easy place to look for your JVM vendor & version, rather than sifting through the comments, so please next time use the template 😃 

Can you try non-Zulu OpenJDK? For example, Temurin (https://adoptium.net/)?

As a further experiment, can you try run it with Java 17 and `-Djdk.net.usePlainSocketImpl`? 
This will revert back to the old Java 12 socket implementation -- trying to verify whether it's an OpenJDK bug.
The old implementation has been removed in Java 18, so you need Java 17 for the system property to have any effect.

**Comment by flozano** [comment]
OK, I have pushed code changes to run the same test with Jetty client, and the results are very different. 
Only with the biggest payload size I'm getting an error:
[failing-jetty-client-4mb.txt](https://github.com/jetty/jetty.project/files/14485692/failing-jetty-client-4mb.txt).
the rest of the cases are fine.

I have tried again with temurin 17 with the system property you mentioned, and even with Semeru JDK 17:
```
2024-03-04T19:36:50.337+0100 [INFO] [org.gradle.process.internal.DefaultExecHandle] Starting process 'Gradle Test Executor 3'. Working directory: /Users/flozano/Projects/java/demo/jetty-issue Command: /Library/Java/JavaVirtualMachines/ibm-semeru-open-17.jdk/Contents/Home/bin/java -Djdk.net.usePlainSocketImpl=true -Dorg.gradle.internal.worker.tmpdir=/Users/flozano/Projects/java/demo/jetty-issue/build/tmp/test/work -Dorg.gradle.native=false @/Users/flozano/.gradle/.tmp/gradle-worker-classpath8841498466969157499txt -Xmx512m -Dfile.encoding=UTF-8 -Duser.country=US -Duser.language=en -Duser.variant -ea worker.org.gradle.process.internal.worker.GradleWorkerMain 'Gradle Test Executor 3'
```
 with same results.
https://github.com/flozano/jetty-issue/commit/cec09fe7805e98714193392fefd35f0fdefd6b57

**Comment by flozano** [comment]
> Had you used the issue template, we would have had an easy place to look for your JVM vendor & version, rather than sifting through the comments, so please next time use the template 😃

I'm very sorry for skipping the process, just edited the initial issue report to follow the template.

**Comment by sbordet** [comment]
@flozano The Jetty client failure you had is because by default, blocking response buffer up to 2 MiB of data, and you are echoing 4 MiB, so:

```
java.lang.IllegalArgumentException: Buffering capacity 2097152 exceeded
```

You can adjust the max buffering size by setting it explicitly, see here how:
https://eclipse.dev/jetty/documentation/jetty-11/programming-guide/index.html#pg-client-http-content-response

I think besides this configuration error, Jetty's `HttpClient` does not have the problem, so it may really be a JDK `HttpClient` bug.

Thanks for your time testing this, BTW.

Let us know if Jetty's `HttpClient` works fine in all cases after the configuration changes.

**Comment by flozano** [comment]
Hi @sbordet , thanks for your help. Yes, I was setting client request/response buffer sizes, but not response listener buffer. I wasn't aware how jetty client works at all.

This makes the tests pass perfectly with jetty client, and therefore it looks like the issue is in JDK client. Initially I discarded this because I had a similar test with undertow and the JDK client, and it worked well... but maybe jetty is writing data faster or in a  different pattern (chunks?) and triggers the bug.

**Comment by sbordet** [comment]
Thanks for the feedback.

Stay with Jetty then 😄

**Comment by joakime** [comment]
> This makes the tests pass perfectly with jetty client, and therefore it looks like the issue is in JDK client. Initially I discarded this because I had a similar test with undertow and the JDK client, and it worked well... but maybe jetty is writing data faster or in a different pattern (chunks?) and triggers the bug.

When the Jetty project itself performs load testing against a Jetty server we have 4+ dedicated Client machines, and 1 server machine.
Anything less, and the server winds up waiting for the clients to catch up.

**Comment by janbartel** [comment]
@flozano are you happy to close this issue as resolved?

**Comment by flozano** [comment]
yes, thank you
